feat(compiler-cli): add extended diagnostics to check for component v…

robertIsaac opened the pull request:
…ariable shadowing template reference

for example
```
@COMPONENT({
  template: '<div #var1></div>'
})
export class FooComponent {
  var1: string;
}
```
fixes #45227

right now this goes without error, but it can lead to mistakenly inside the template to use the template reference as if it was the component variable

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 45227


## What is the new behavior?
it will show warning when there is shadowing between component variable and template reference


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

I guess it's not, because it's a warning not an error, but I'm leaving it for the Angular team to decide

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
to fix the warning, simply rename the component variable or the template reference

## Other information
this will show the warning even if the component variable is private, which in my opinion is correct
but in case the Angular team disagree, I will make it only show the warning in case of protect or public variables only

robertIsaac committed:
feat(compiler-cli): add extended diagnostics to check for component variable shadowing template reference

for example
```
@COMPONENT({
  template: '<div #var1></div>'
})
export class FooComponent {
  var1: string;
}
```

right now this goes without error, but it can lead to mistakenly inside the template to use the template reference as if it was the component variable

Fixes #45227

JoostK reviewed (changes requested):
There's a few more cases that such check would need to take into account:

1. Inherited properties will not cause the message to be reported.
2. Template scopes can shadow ancestor template variables:

```html
<div *ngFor="let local of []">
  <button #local></button>
</div>
```

Here, `#local` shadows the `let-local` declaration of the `ng-template` element that is created by the structural `ngFor` directive.

AndrewKushnir reviewed (commented):
@robertIsaac thanks for the PR 👍

I've left some comments, including additional information about build failures from the internal tests run.

robertIsaac commented on packages/compiler-cli/src/ngtsc/typecheck/extended/checks/component_variable_shadow_template_reference/index.ts:
```diff
@@ -0,0 +1,54 @@
+ */
+
+import {AST, TmplAstNode} from '@angular/compiler';
+import {Reference} from '@angular/compiler/src/render3/r3_ast';
```
done

robertIsaac replied:
I see there is more imports from `src/render3/r3_ast`, should I fix them as well?

robertIsaac commented on packages/compiler-cli/src/ngtsc/typecheck/extended/checks/component_variable_shadow_template_reference/index.ts:
```diff
@@ -0,0 +1,54 @@
+
+import {AST, TmplAstNode} from '@angular/compiler';
+import {Reference} from '@angular/compiler/src/render3/r3_ast';
+import ts, {Identifier} from 'typescript';
```
done

robertIsaac commented on packages/compiler-cli/src/ngtsc/typecheck/extended/checks/component_variable_shadow_template_reference/index.ts:
```diff
@@ -0,0 +1,54 @@
+      ): NgTemplateDiagnostic<ErrorCode.COMPONENT_VARIABLE_SHADOW_TEMPLATE_REFERENCE>[] {
+    if (!(node instanceof Reference)) return [];
+
+    const templateReference = node.name;
```
do I need to explicitly provide type to `member` or what? because in my IDE it shows correctly `member: ts.ClassElement`

robertIsaac commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -318,6 +318,21 @@ export enum ErrorCode {
   OPTIONAL_CHAIN_NOT_NULLABLE = 8107,
 
+  /**
+   * A variable in the component shadow the same name as a template reference.
```
done

robertIsaac commented on goldens/public-api/compiler-cli/extended_template_diagnostic_name.md:
```diff
@@ -6,6 +6,8 @@
 // @public
 export enum ExtendedTemplateDiagnosticName {
+    // (undocumented)
+    COMPONENT_VARIABLE_SHADOW_TEMPLATE_REFERENCE = "ComponentVariableShadowTemplateReference",
```
I renamed everything to be shadows

robertIsaac commented:
> 1. Inherited properties will not cause the message to be reported.

so members of the component return only direct members? is there anyway to get all the members of a class?

> 2. Template scopes can shadow ancestor template variables:
> 
> ```
> <div *ngFor="let local of []">
>   <button #local></button>
> </div>
> ```
> 
> Here, `#local` shadows the `let-local` declaration of the `ng-template` element that is created by the structural `ngFor` directive.

oh I missed this when I was solving the solution, will see what I can do about it tomorrow

robertIsaac commented:
@JoostK I found making `componentVariables` to equal `(component.parent as unknown as {identifiers: Map<string, string>}).identifiers.keys();` solves the first problem
but there is two problems
1. the type of parent is node which shouldn't have identifiers
2. it will also forbid using the class names of the component and its superclasses
any suggestions here?

JoostK commented:
> so members of the component return only direct members? is there anyway to get all the members of a class?

Indeed, `members` is a property of the parse tree so it corresponds with syntactical items of the class. You'll need a `ts.TypeChecker` to obtain all properties for a given declaration (I'm on mobile atm and don't know how to get to that, not which methods to then use, by heart)

robertIsaac committed:
feat(compiler-cli): add extended diagnostics to check for component variable shadowing template reference

for example
```
@COMPONENT({
  template: '<div #var1></div>'
})
export class FooComponent {
  var1: string;
}
```

right now this goes without error, but it can lead to mistakenly inside the template to use the template reference as if it was the component variable

Fixes #45227

robertIsaac commented:
@JoostK I made test cases to check how actually angular will work with edge cases
https://stackblitz.com/edit/angular-ivy-jnn75s

and it seems things are actually more complicated than what I originally thought
the template reference has scope for structural directives tree, but if there is none then it's global for the template

so the acceptance criteria from my opinion is:
1. any template reference shouldn't shadow any component variable (that's done in this PR)
2. any two template references can shadow each other as long as they have different scopes - inside different structural directive elements (that's the current behavior and we need to maintain it)
3. scoped references shouldn't shallow global ones (not done)
4. inside same scope, no two reference should shallow each other (not done)

do you think there is more edge cases I should consider?

robertIsaac commented:
@AndrewKushnir @JoostK any update about this PR?

JeanMeche commented:
Can you rebase your branch to fix the conflicts ?

robertIsaac commented:
> Can you rebase your branch to fix the conflicts ? (Error 8108 is now taken).

done

do you plan to accept it?

jessicajaniuk commented:
@robertIsaac looks like you have failing status checks. You'll need to resolve those first. One of them is showing that you need to update goldens now that you've added this check.

JeanMeche commented:
(You'll need to run `yarn bazel run //packages/compiler-cli:error_code_api.accept` for the tests to pass).

robertIsaac commented:
> (You'll need to run `yarn bazel run //packages/compiler-cli:error_code_api.accept` for the tests to pass).

done

JoostK commented:
We discussed this just now and decided to close this for now, as the PR in its current state is not ready to land:

1. it uses internal fields of TS AST nodes, which we'd rather avoid as it's brittle w.r.t TS updates, as well as being incomplete w.r.t. inherited fields.
2. There's no shadowing checks for nested templates (which may shadow variable declarations of enclosing templates)

robertIsaac commented:
hi @JoostK thanks for your feedback
but I'm interested to know what should be the proper way to do it
in a previous comment it was mentioned I should use `ts.TypeChecker` but I never understood how exactly to use it

JoostK commented:
`ts.TypeChecker` provides a method `getPropertiesOfType` to go from a `ts.Type` to its properties as `ts.Symbol`. Those symbols also have names, and that result also includes properties from base classes, which wouldn't be part of the class' AST structure. An extended diagnostic has access to the type checker in `ctx.typeChecker` (with `ctx` being `TemplateContext` that is passed as first parameter of `TemplateCheck.run`).

You'd also have to go from the `ts.ClassDeclaration` to its `ts.Type` using `ts.TypeChecker.getTypeAtLocation()`, arriving at something like the following:

```ts
const properties = ctx.typeChecker.getPropertiesOfType(ctx.typeChecker.getTypeAtLocation(component));
```
